python/tvm/target/detect_target.py

```diff
@@ -0,0 +1,169 @@
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Detect target."""
+from typing import Union
+
+from . import Target
+from .._ffi import get_global_func
+from .._ffi.runtime_ctypes import Device
+from ..runtime.ndarray import device
+
+
+def _detect_metal_host() -> "Target":
+    target_triple = get_global_func("tvm.codegen.llvm.GetDefaultTargetTriple")()
+    process_triple = get_global_func("tvm.codegen.llvm.GetProcessTriple")()
+    host_cpu = get_global_func("tvm.codegen.llvm.GetHostCPUName")()
+    print(
+        f"Host CPU dection:\n  Target triple: {target_triple}\n  "
+        f"Process triple: {process_triple}\n  Host CPU: {host_cpu}"
+    )
+    if target_triple.startswith("x86_64-"):
+        return Target(
+            {
+                "kind": "llvm",
+                "mtriple": "x86_64-apple-macos",
+                "mcpu": host_cpu,
+            }
+        )
+    # should start with "arm64-"
+    return Target(
+        {
+            "kind": "llvm",
+            "mtriple": "arm64-apple-macos",
+            "mcpu": host_cpu,
+        }
+    )
+
+
+def _detect_metal(dev: Device) -> Target:
+    return Target(
+        {
+            "kind": "metal",
+            "max_shared_memory_per_block": 32768,
+            "max_threads_per_block": dev.max_threads_per_block,
+            "thread_warp_size": 32,
+        },
+        host=_detect_metal_host(),
+    )
+
+
+def _detect_cuda(dev: Device) -> Target:
+    return Target(
+        {
+            "kind": "cuda",
+            "max_shared_memory_per_block": dev.max_shared_memory_per_block,
+            "max_threads_per_block": dev.max_threads_per_block,
+            "thread_warp_size": dev.warp_size,
+            "registers_per_block": 65536,
+            "arch": "sm_" + dev.compute_version.replace(".", ""),
+        }
+    )
+
+
+def _detect_rocm(dev: Device) -> Target:
+    return Target(
+        {
+            "kind": "rocm",
+            "max_shared_memory_per_block": dev.max_shared_memory_per_block,
+            "max_threads_per_block": dev.max_threads_per_block,
+            "thread_warp_size": dev.warp_size,
+        }
+    )
+
+
+def _detect_vulkan(dev: Device) -> Target:
+    return Target(
+        {
+            "kind": "vulkan",
+            "max_threads_per_block": dev.max_threads_per_block,
+            "max_shared_memory_per_block": dev.max_shared_memory_per_block,
+            "thread_warp_size": dev.warp_size,
+            "supports_float16": 1,
+            "supports_int16": 1,
+            "supports_int8": 1,
+            "supports_16bit_buffer": 1,
+        }
+    )
+
+
+def _detect_opencl(dev: Device) -> Target:  # pylint: disable=unused-argument
+    return Target("opencl")
+
+
+def _detect_cpu(dev: Device) -> Target:  # pylint: disable=unused-argument
+    return Target("llvm")
```

Comment:
Let's remove those two because auto detect doesn't seem to be properly supported yet